Core/HLE/sceNet.cpp

```diff
@@ -191,11 +233,26 @@ void __NetDoState(PointerWrap &p) {
 		p.Do(netThread1Addr);
 		p.Do(netThread2Addr);
 	}
+	if (s >= 4) {
+		p.Do(netApctlState);
+		p.Do(netApctlInfo);
+		p.Do(actionAfterApctlMipsCall);
+		__KernelRestoreActionType(actionAfterApctlMipsCall, AfterApctlMipsCall::Create);
+		p.Do(apctlThreadHackAddr);
+	}
 	// Let's not change "Inited" value when Loading SaveState in the middle of multiplayer to prevent memory & port leaks
 	if (p.mode == p.MODE_READ) {
 		netApctlInited = cur_netApctlInited;
 		netInetInited = cur_netInetInited;
 		netInited = cur_netInited;
+
+		// Previously, this wasn't being saved.  It needs its own space.
+		if (apctlThreadHackAddr && strcmp("apctlThreadHack", kernelMemory.GetBlockTag(apctlThreadHackAddr)) != 0) {
+			u32 blockSize = sizeof(apctlThreadCode);
+			apctlThreadHackAddr = kernelMemory.Alloc(blockSize, false, "apctlThreadHack");
+		}
+		// Restore Apctl Loop MIPS code to prevent crashes after loading from SaveState
+		if (apctlThreadHackAddr) Memory::Memcpy(apctlThreadHackAddr, apctlThreadCode, sizeof(apctlThreadCode));
```

Comment:
This is also a confusing part for me, as i remembered when loading save state memory got reset thus the loops mipscall no longer exist and could lead to crashes.

But if i recreate the loop code it here, it will cause AfterAction goes to the wrong PSPAction when loading from old savestate, for example in this case `AfterMatchingMipsCall::run` is triggered instead of  `AfterApctlMipsCall::run` when i checked the Debug Console Log. 

The same issue also happened when the first time i implements `AfterAdhocMipsCall` which also triggers `AfterMatchingMipsCall::run` when loaded from old savestate that have no `AfterAdhocMipsCall` yet.

Reply:
You might want to just reject old save states by incrementing the first number, in case loading those savestates no longer makes sense.

In fact I'm wondering how it makes sense to use save states at all during multiplayer ... I guess if you manually do a synchronized load on all playing devices there's a chance to recover sync?

Reply:
The important thing to ask is: what happens if I'm not using multiplayer right now?

Example: suppose I'm playing Yu Gi Oh or something.  I do a network battle (I think it has those.)  Then I go on with my single-player life.  Later, I save state.

Will any of the network code be involved in this save state now?  If yes, we shouldn't reject loading this save state and should try to keep backwards compatibility (at least being able to load) if possible.

-[Unknown]

Reply:
After testing it further this is the situation:
1). When the first time trying multiplayer, `actionAfterApctlMipsCall` have a value of 9 and `actionAfterMatchingMipsCall` have a value of 10
2). When loading old save state that doesn't have `actionAfterApctlMipsCall` yet, the `actionAfterApctlMipsCall` remains the same (9) but `actionAfterMatchingMipsCall` became 9 also, so this is probably why the Apctl callbacks is triggering PSPAction of Matching callbacks instead.